store/tikv/region_request.go

```diff
@@ -633,14 +653,22 @@ func regionErrorToLabel(e *errorpb.Error) string {
 	return "unknown"
 }
 
-func (s *RegionRequestSender) onRegionError(bo *Backoffer, ctx *RPCContext, seed *uint32, regionErr *errorpb.Error) (shouldRetry bool, err error) {
+func (s *RegionRequestSender) onRegionError(bo *Backoffer, ctx *RPCContext, seed *uint32, regionErr *errorpb.Error) (shouldRetry bool, opts []StoreSelectorOption, err error) {
 	if span := opentracing.SpanFromContext(bo.GetCtx()); span != nil && span.Tracer() != nil {
 		span1 := span.Tracer().StartSpan("tikv.onRegionError", opentracing.ChildOf(span.Context()))
 		defer span1.Finish()
 		// TODO(MyonKeminta): Make sure trace works without cloning the backoffer.
 		// bo = bo.Clone()
 		bo.SetCtx(opentracing.ContextWithSpan(bo.GetCtx(), span1))
 	}
+	opts = ctx.storeSelectorOptions
+	// Stale Read request will retry the leader or next peer on error,
+	// so we will exclude the StoreID of the requested peer every time.
+	// If the new StoreID keeps being the same with the last one, we
+	// should not continue excluding it to make the opts become bigger.
+	if ctx.Store != nil && ctx.isStaleRead && ctx.lastStoreID != ctx.Store.storeID {
```

Comment:
Do you want to retry the next peer with the opts? It may handle it same as https://github.com/pingcap/tidb/blob/ca3d88eba5c67c17ed0f56faa10dcef8af6236d3/store/tikv/region_cache.go#L844-L850 in the `sendReqToRegion`.

Reply:
This part is used to handle the internal RPC error and a stale read request may also trigger it when an RPC error occurs. This PR mainly wants to handle the error in a normal RPC response and retry on different peers by excluding stores by `opts`.

Reply:
Yes, there is another way with the seed:
https://github.com/pingcap/tidb/blob/8656b5d39668fb1511bc2cdaecf7f6ccb757c9bd/store/tikv/region_request.go#L690-L698.

BTW, Will the retry cause the cross-region issue in cross-region deployment?

Reply:
You're right. I have changed to use the `seed` increment to try the next peer, and it works well!

Retry won't make the request become cross-region because while retrying the `WithMatchLabels` will keep filtering the stores located in different DCs.

https://github.com/pingcap/tidb/blob/9fad132d1d892ade178d0614687660c29cc5af54/store/tikv/snapshot.go#L475-L483